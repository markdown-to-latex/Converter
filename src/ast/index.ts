export * as node from './node';
export * as parsing from './parsing';
export * from './nodes';
