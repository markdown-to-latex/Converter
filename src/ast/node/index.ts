export * from './extension';
export * from './function';
export * from './position';
export * from './struct';
