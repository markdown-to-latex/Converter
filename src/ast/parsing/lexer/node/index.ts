export * as blockquote from './blockquote';
export * as breaks from './breaks';
export * as code from './code';
export * as codeSpan from './codeSpan';
export * as comment from './comment';
export * as del from './del';
export * as em from './em';
export * as formulaOrLatex from './formulaOrLatex';
export * as formulaSpan from './formulaSpan';
export * as heading from './heading';
export * as hr from './hr';
export * as image from './image';
export * as latexSpan from './latexSpan';
export * as link from './link';
export * as list from './list';
export * as macros from './macros';
export * as space from './space';
export * as strong from './strong';
export * as table from './table';
export * as underline from './underline';
