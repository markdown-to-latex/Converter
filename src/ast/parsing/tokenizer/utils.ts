export const LINE_SPLIT_REGEXP = new RegExp(/\r?\n/g);

export function stringHexlify(str: string): string{
    let hex, i;

    let result = "";
    for (i=0; i<str.length; i++) {
        hex = str.charCodeAt(i).toString(16);
        result += ("000"+hex).slice(-4);
    }

    return result
}
