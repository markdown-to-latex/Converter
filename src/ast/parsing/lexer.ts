import * as node from '../node';
import {
    CodeNode,
    copyTextPosition,
    createStartEndPos,
    Node,
    NodeType,
    RawNode,
    RawNodeE,
    RawNodeType,
    TextPosition,
    textPositionG,
    textPositionGEq,
} from '../node';
import {
    DiagnoseErrorType,
    DiagnoseList,
    diagnoseListHasSeverity,
    DiagnoseSeverity,
} from '../../diagnose';

export const enum LexemeType {
    // Space,
    // Paragraph,
    // Heading,
    // Text,
    Code,
}

const parserPriorities: LexemeType[] = [
    LexemeType.Code,
    // LexemeType.Heading,
    // LexemeType.Paragraph,
    // LexemeType.Text,
    // LexemeType.Space,
];

interface LexemeTypeToNodeType {
    [LexemeType.Code]: node.CodeNode;
    // [LexemeType.Heading]: node.HeadingNode;
    // [LexemeType.Paragraph]: node.ParagraphNode;
    // [LexemeType.Text]: node.TextNode;
    // [LexemeType.Space]: node.SpaceNode;
}

class FatalError extends Error {
}

export function applyVisitors(nodes: Readonly<Node[]>): [Node[], DiagnoseList] {
    let allDiags: DiagnoseList = [];

    let newNodes = [...nodes];
    try {
        for (const type of parserPriorities) {
            newNodes = newNodes.flatMap(node => {
                if (node.type !== RawNodeType.Raw) {
                    return [node];
                }

                let [nodes, diags] = applyParser(node as RawNode, type);
                allDiags.push(...diags);

                if (diagnoseListHasSeverity(diags, DiagnoseSeverity.Fatal)) {
                    throw new FatalError();
                }

                return nodes;
            });
        }
    } catch (e) {
        if (e instanceof FatalError) {
            return [[], allDiags];
        } else {
            throw e;
        }
    }

    return [newNodes, allDiags];
}

function applyParser(
    node: Readonly<RawNode>,
    type: LexemeType,
): [Node[], DiagnoseList] {
    const nodeE = new RawNodeE(node);

    const diags: DiagnoseList = [];
    const newNodes: Node[] = [];
    let position: TextPosition = copyTextPosition(node.pos.start);

    const parsed = parsers[type](nodeE);
    for (const node of parsed) {
        if (!textPositionGEq(position, node.pos.start)) {
            newNodes.push({
                type: RawNodeType.Raw,
                pos: {
                    start: copyTextPosition(position),
                    end: copyTextPosition(node.pos.start),
                },
                text: nodeE.text.slicePosition(
                    nodeE.n.pos.start,
                    position,
                    node.pos.start,
                ).s,
                parent: nodeE.n.parent,
            } as RawNode);
        }

        newNodes.push(node);
        position = copyTextPosition(node.pos.end);
    }

    if (!textPositionGEq(position, node.pos.end)) {
        newNodes.push({
            type: RawNodeType.Raw,
            pos: {
                start: copyTextPosition(position),
                end: copyTextPosition(node.pos.end),
            },
            text: nodeE.text.slicePosition(
                nodeE.n.pos.start,
                position,
                node.pos.end,
            ).s,
            parent: nodeE.n.parent,
        } as RawNode);
    } else if (textPositionG(position, node.pos.end)) {
        diags.push({
            errorType: DiagnoseErrorType.ApplyParserError,
            severity: DiagnoseSeverity.Fatal,
            pos: {
                start: node.pos.end,
                end: position,
            },
            message: 'Parser position overflow',
            filePath: nodeE.parentFile?.path ?? 'null',
        });
    }

    return [newNodes, diags];
}

type Visitor<T> = (node: RawNodeE) => T[];
const parsers: {
    [key in LexemeType]: Visitor<LexemeTypeToNodeType[key]>;
} = {
    [LexemeType.Code]: node => {
        const lines = node.text.lines;

        const codeLexemes = lines
            .map((v, i) => ({
                offLine: i,
                absLine: node.n.pos.start.line + i,
                match: v.s.match(/\s*`{3,}/),
            }))
            .filter(v => v.match?.length);
        if (codeLexemes.length % 2 !== 0) {
            const lastLexeme = codeLexemes[codeLexemes.length - 1];
            const absLine = lastLexeme.absLine;
            throw new Error(
                `Unable to find closing quotes for block code. ` +
                `Began at ${absLine} at file TODO`,
            );
        }

        const codeNodes: CodeNode[] = [];
        for (let i = 0; i < codeLexemes.length; i += 2) {
            const startLexeme = codeLexemes[i];
            const endLexeme = codeLexemes[i + 1];

            const text = lines
                .slice(startLexeme.offLine + 1, endLexeme.offLine)
                .join('\n');
            codeNodes.push({
                type: NodeType.Code,
                pos: createStartEndPos(
                    startLexeme.absLine,
                    1,
                    endLexeme.absLine,
                    lines[endLexeme.offLine].length,
                ),
                text: text,
                parent: node.n.parent,
                lang: null,
                codeBlockStyle: null,
            });
        }

        return codeNodes;
    },
};
