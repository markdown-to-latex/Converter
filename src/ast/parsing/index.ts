import {
    createStartEndTextPos,
    FileNode,
    NodeE,
    NodeType,
    RawNode,
    RawNodeType,
    StartEndTextPosition,
} from '../node';
import path from 'path';
import { StringE } from '../../extension/string';
import { tokenize, tokensToNode } from './tokenizer';

export function fullContentPos(
    content: string | StringE,
): StartEndTextPosition {
    const contentE = StringE.from(content);

    const lines = contentE.splitE(/\r?\n/);
    return createStartEndTextPos(
        1,
        1,
        lines.length,
        lines[lines.length - 1].length + 1,
    );
}

function parseFile(content: string, filePath: string) {
    const contentNode: RawNode = {
        type: RawNodeType.Raw,
        parent: null,
        text: content,
        pos: {
            start: 0,
            end: content.length,
        },
    };

    const fileNode: FileNode = {
        type: NodeType.File,
        parent: null,
        children: [contentNode],
        pos: {
            start: 0,
            end: content.length,
        },
        raw: content,
        path: path.resolve(filePath),
    };

    contentNode.parent = fileNode;

    let nodeE = NodeE.from(fileNode);
}
