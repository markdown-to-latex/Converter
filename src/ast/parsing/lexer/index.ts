import * as node from "../../node";
import {
    getNodeParentFile,
    Node,
    NodeChildren,
    NodeText,
    NodeType,
    positionToTextPosition,
    RawNode,
    RawNodeType,
    TextNode,
    TokensNode
} from "../../node";
import {
    DiagnoseErrorType,
    DiagnoseInfo,
    DiagnoseList,
    diagnoseListHasSeverity,
    DiagnoseSeverity
} from "../../../diagnose";
import { Token, tokenize, tokensToNode, TokenType } from "../tokenizer";
import { TokenByTypeParserResult, TokenParser, TokenPredicate } from "./struct";
import { parseCode } from "./node/code";
import {
    isParagraphBreak,
    parseParagraphBreak,
    parseSoftBreak
} from "./node/paragraph";
import { parseCodeSpan } from "./node/codeSpan";
import { parseLink } from "./node/link";
import { parseMacro } from "./node/macros";
import { parseTable } from "./node/table";
import { parseList } from "./node/list";
import { parseHeading } from "./node/heading";
import { parseBlockquote } from "./node/blockquote";
import { parseImage } from "./node/image";
import { parseHr } from "./node/hr";
import { parseEm } from "./node/em";
import { parseStrongWithOptionalEm } from "./node/strong";

export const enum LexemeType {
    // Space,
    // Paragraph,
    // Heading,
    // Text,
    Code,
}

const parserPriorities: LexemeType[] = [
    LexemeType.Code
    // LexemeType.Heading,
    // LexemeType.Paragraph,
    // LexemeType.Text,
    // LexemeType.Space,
];

interface LexemeTypeToNodeType {
    [LexemeType.Code]: node.CodeNode;
    // [LexemeType.Heading]: node.HeadingNode;
    // [LexemeType.Paragraph]: node.ParagraphNode;
    // [LexemeType.Text]: node.TextNode;
    // [LexemeType.Space]: node.SpaceNode;
}

class FatalError extends Error {
}

interface ApplyVisitorsResult {
    nodes: Node[];
    diagnostic: DiagnoseList;
}

export function applyVisitors(nodes: Readonly<Node[]>): ApplyVisitorsResult {
    const visitorsResult: ApplyVisitorsResult = {
        nodes: [...nodes],
        diagnostic: []
    };

    try {
        for (const type of parserPriorities) {
            visitorsResult.nodes = visitorsResult.nodes.flatMap(node => {
                if (node.type === RawNodeType.Raw) {
                    const parent = node.parent;
                    node = tokensToNode(tokenize((node as RawNode).text, 0));
                    node.parent = parent;
                }

                if (node.type !== RawNodeType.Tokens) {
                    return [node];
                }

                const result = parseTokensNode(node as TokensNode);
                visitorsResult.diagnostic.push(...result.diagnostic);

                if (
                    diagnoseListHasSeverity(
                        result.diagnostic,
                        DiagnoseSeverity.Fatal
                    )
                ) {
                    throw new FatalError();
                }

                return result.nodes;
            });
        }
    } catch (e) {
        if (e instanceof FatalError) {
            return visitorsResult;
        } else {
            throw e;
        }
    }

    return visitorsResult;
}

interface FindTokenResult {
    token: Token;
    index: number;
}

export function findTokenOrNull(
    tokens: TokensNode,
    fromIndex: number,
    predicate: TokenPredicate
): FindTokenResult | null {
    for (let i = fromIndex; i < tokens.tokens.length; ++i) {
        const token = tokens.tokens[i];
        if (predicate(token, i, tokens)) {
            return {
                token,
                index: i
            };
        }
    }
    return null;
}

export function findTokenOrNullBackward(
    tokens: TokensNode,
    fromIndex: number,
    predicate: TokenPredicate
): FindTokenResult | null {
    for (let i = fromIndex; i >= 0; --i) {
        const token = tokens.tokens[i];
        if (predicate(token, i, tokens)) {
            return {
                token,
                index: i
            };
        }
    }
    return null;
}

const brackets: Record<string, string> = {
    "[": "]", // Label
    "{": "}", // Reserved
    "(": ")", // Argument
    "`": "`", // Code Span
    "$`": "`$" // Formula Span
};

const bracketsBackward: Record<string, string> = Object.fromEntries(
    Object.entries(brackets).map(([k, v]) => [v, k])
);

const exclusiveBrackets = ["`", "$`"];

export function findTokenClosingBracket(
    tokens: TokensNode,
    index: number,
    greedy: boolean = false /* Capture self-closing bracket only */
): FindTokenResult | null {
    const openToken = tokens.tokens[index];
    const closeBracket = brackets[openToken.text] ?? openToken.text;

    const bracketsCounter: Record<string, number> = Object.fromEntries(
        Object.keys(brackets).map(k => [k, 0])
    );

    for (let i = index + 1; i < tokens.tokens.length; ++i) {
        const token = tokens.tokens[i];
        if (isParagraphBreak(token, i, tokens)) {
            return null; // break breaks the brackets
        }

        if (
            token.type !== TokenType.JoinableSpecial &&
            token.type !== TokenType.SeparatedSpecial
        ) {
            continue;
        }

        if (greedy) {
            if (token.text === closeBracket) {
                return {
                    token: token,
                    index: i
                };
            }
        } else {
            if (
                Object.values(bracketsCounter).reduce(
                    (prev, current) => prev + current,
                    0
                ) === 0
            ) {
                if (token.text === closeBracket) {
                    return {
                        token: token,
                        index: i
                    };
                }
            }

            const exclusiveBracket =
                exclusiveBrackets
                    .map(v => [v, bracketsCounter[v]] as const)
                    .find(v => v[1]) ?? null;
            if (exclusiveBracket) {
                const exclusiveOpenBracket = exclusiveBracket[0];
                const exclusiveCloseBracket = brackets[exclusiveOpenBracket];
                if (token.text == exclusiveCloseBracket) {
                    bracketsCounter[exclusiveOpenBracket] -= 1;
                }
            } else {
                // Count all inner brackets
                if (token.text in brackets) {
                    bracketsCounter[token.text] += 1;
                } else if (token.text in bracketsBackward) {
                    bracketsCounter[bracketsBackward[token.text]] -= 1;
                }
            }
        }
    }

    return null;
}

export function isOpenLabelBracket(tokens: TokensNode, index: number): boolean {
    const token = tokens.tokens[index];

    if (
        token.type !== TokenType.JoinableSpecial &&
        token.type !== TokenType.SeparatedSpecial
    ) {
        return false;
    }

    return token.text == "[";
}

export function isOpenArgumentBracket(
    tokens: TokensNode,
    index: number
): boolean {
    const token = tokens.tokens[index];

    if (
        token.type !== TokenType.JoinableSpecial &&
        token.type !== TokenType.SeparatedSpecial
    ) {
        return false;
    }

    return token.text == "(";
}

interface TrimWithPositionsResult {
    result: string;
    leftTrim: number;
    rightTrim: number;
}

export function trimSingleWithPositions(str: string): TrimWithPositionsResult {
    const originalLen = str.length;
    const leftLen = originalLen - str.trimStart().length;
    const rightLen = originalLen - str.trimEnd().length;

    if (originalLen <= leftLen + rightLen) {
        return {
            result: " ",
            leftTrim: originalLen - 1,
            rightTrim: 0
        };
    }

    return {
        result: str.slice(
            leftLen > 0 ? leftLen - 1 : 0,
            rightLen > 0 ? originalLen - rightLen + 1 : originalLen
        ),
        leftTrim: leftLen > 0 ? leftLen - 1 : 0,
        rightTrim: rightLen > 0 ? rightLen - 1 : 0
    };
}

interface ParseTokensNodeResult {
    nodes: Node[];
    diagnostic: DiagnoseList;
}

export function parseTokensNode(tokens: TokensNode): ParseTokensNodeResult {
    const parsingResult: ParseTokensNodeResult = {
        nodes: [],
        diagnostic: []
    };

    for (let i = 0; i < tokens.tokens.length /*manual*/;) {
        const token = tokens.tokens[i];

        const result = parseTokensNodeByType(tokens, i);
        if (result) {
            i = result.index;
            parsingResult.diagnostic.push(...result.diagnostic);
            parsingResult.nodes.push(...result.nodes);

            if (
                diagnoseListHasSeverity(
                    result.diagnostic,
                    DiagnoseSeverity.Fatal
                )
            ) {
                return parsingResult;
            }

            continue;
        }

        parsingResult.nodes.push({
            type: NodeType.Text,
            parent: tokens.parent,
            text: token.text,
            pos: {
                start: token.pos,
                end: token.pos + token.text.length
            },
            children: []
        } as TextNode);
        ++i;
    }

    nodeJoiner(parsingResult.nodes);
    // parsingResult.nodes = parsingResult.nodes.map(n => {
    //     if (n.type === NodeType.Text) {
    //         const trimmed = trimSingleWithPositions((n as TextNode).text);
    //         return {
    //             type: NodeType.Text,
    //             text: trimmed.result,
    //             parent: n.parent,
    //             pos: {
    //                 start: n.pos.start + trimmed.leftTrim,
    //                 end: n.pos.end - trimmed.rightTrim,
    //             },
    //             children: (n as TextNode).children,
    //         } as TextNode;
    //     }
    //     return n;
    // });
    parsingResult.nodes = parsingResult.nodes.filter(
        n => n.type !== RawNodeType.ParagraphBreak
    );
    return parsingResult;
}

function nodeJoiner(nodes: Node[]): void {
    const breaks: (NodeType | RawNodeType)[] = [
        RawNodeType.SoftBreak,
        RawNodeType.ParagraphBreak
    ];
    if (breaks.indexOf(nodes[0]?.type) !== -1) {
        nodes.splice(0, 1);
    }

    // TODO: Trim nodes
    while (nodes.length) {
        const textNode = nodes[0] as TextNode;

        if (
            textNode.type !== NodeType.Text ||
            textNode.text.trim().length !== 0
        ) {
            break;
        }

        nodes.splice(0, 1);
    }
    while (nodes.length) {
        const textNode = nodes[nodes.length - 1] as TextNode;

        if (
            textNode.type !== NodeType.Text ||
            textNode.text.trim().length !== 0
        ) {
            break;
        }

        nodes.splice(nodes.length - 1, 1);
    }

    let index = 1;
    while (index < nodes.length) {
        const currentNode = nodes[index];
        const prevNode = nodes[index - 1];

        if (currentNode.type === RawNodeType.SoftBreak) {
            if (
                prevNode.type === NodeType.Text &&
                nodes[index + 1]?.type === NodeType.Text
            ) {
                const prevNodeText = prevNode as Node & NodeText;

                prevNodeText.pos.end = currentNode.pos.end;
                prevNodeText.text += " ";
            }

            nodes.splice(index, 1);
            continue;
        }

        if (
            currentNode.type === prevNode.type &&
            currentNode.type === NodeType.Text
        ) {
            // Contains 'text' field => is a NodeText
            const currentNodeText = currentNode as Node & NodeText;
            const prevNodeText = prevNode as Node & NodeText;

            prevNodeText.pos.end = currentNodeText.pos.end;
            prevNodeText.text += currentNodeText.text;

            nodes.splice(index, 1);
            continue;
        }

        if (
            currentNode.type === prevNode.type &&
            currentNode.type === NodeType.Blockquote
        ) {
            // Contains 'text' field => is a NodeText
            const currentNodeChildren = currentNode as Node & NodeChildren;
            const prevNodeChildren = prevNode as Node & NodeChildren;

            prevNodeChildren.pos.end = currentNodeChildren.pos.end;
            prevNodeChildren.children.push(...currentNodeChildren.children);
            currentNodeChildren.children.forEach(
                n => (n.parent = prevNodeChildren)
            );

            nodes.splice(index, 1);
            continue;
        }

        ++index;
    }
}

function parseTokensNodeByType(
    tokens: TokensNode,
    index: number
): TokenByTypeParserResult | null {
    const token = tokens.tokens[index];

    for (const parser of parsersByType[token.type]) {
        const result = parser(tokens, index);
        if (result) {
            return result;
        }
    }

    return null;
}

const parsersByType: Record<TokenType, TokenParser[]> = {
    [TokenType.JoinableSpecial]: [
        parseCodeSpan,
        parseStrongWithOptionalEm,
        parseEm,
        parseHeading,
        parseHr,
        parseCode
    ],
    [TokenType.SeparatedSpecial]: [
        parseLink,
        parseMacro,
        parseImage,
        parseTable,
        parseBlockquote
    ],
    [TokenType.Delimiter]: [parseSoftBreak, parseParagraphBreak],
    [TokenType.Spacer]: [parseList],
    [TokenType.Letter]: [parseList],
    [TokenType.Other]: []
};

export function tokenToDiagnose(
    tokens: TokensNode,
    index: number,
    message: string,
    severity: DiagnoseSeverity = DiagnoseSeverity.Fatal
): DiagnoseInfo {
    const token = tokens.tokens[index];
    const nodeParentFile = getNodeParentFile(tokens);
    return {
        message: message,
        severity: severity,
        filePath: nodeParentFile?.path ?? "null",
        errorType: DiagnoseErrorType.ApplyParserError,
        pos: {
            start: {
                absolute: token.pos,
                ...positionToTextPosition(nodeParentFile?.raw ?? "", token.pos)
            },
            end: {
                absolute: token.pos + token.text.length,
                ...positionToTextPosition(
                    nodeParentFile?.raw ?? "",
                    token.pos + token.text.length
                )
            }
        }
    };
}

export function unexpectedEof(
    tokens: TokensNode,
    index: number,
    message: string
): TokenByTypeParserResult {
    return {
        nodes: [],
        index: tokens.tokens.length,
        diagnostic: [
            tokenToDiagnose(tokens, index, message, DiagnoseSeverity.Error)
        ]
    };
}

export function sliceTokenText(
    tokens: TokensNode,
    fromIndex: number,
    toIndex: number
): string {
    const tokenStart = tokens.tokens[fromIndex];
    const tokenEnd: Token | null = tokens.tokens[toIndex];

    return tokens.text.slice(
        tokenStart.pos - tokens.pos.start,
        tokenEnd ? tokenEnd.pos - tokens.pos.end : tokens.pos.end
    );
}
