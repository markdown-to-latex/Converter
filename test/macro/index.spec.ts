import {rawNodeTemplate} from "../ast/parsing/lexer/utils";
import {applyVisitors} from "../../src/ast/parsing/lexer";
import {FileNode, ParagraphNode} from "../../src/ast/node";
import {applyMacros} from "../../src/macro";

describe('macros', () => {
    test('table', () => {
        const rawNode = rawNodeTemplate(`
!TK[table-label]

!T[table-label](a table)

|asda|asdasd|
|----|------|
|asda|asdasd|

`);
        const { nodes } = applyVisitors([rawNode]);
        (rawNode.parent as FileNode).children = nodes;

        const diagnostic = applyMacros(rawNode.parent as FileNode);
        expect(diagnostic).toHaveLength(0)
        // TODO: not working, needs in-depth traverse
        expect(nodes).toMatchSnapshot();
    })
});