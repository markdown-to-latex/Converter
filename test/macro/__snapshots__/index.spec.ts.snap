// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros table 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "parent": [Circular],
        "pos": Object {
          "end": 17,
          "start": 1,
        },
        "text": "0",
        "type": "Text",
      },
      Object {
        "parent": [Circular],
        "pos": Object {
          "end": 26,
          "start": 18,
        },
        "type": "Br",
      },
    ],
    "parent": Object {
      "children": [Circular],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 96,
        "start": 0,
      },
      "raw": "
!TK[table-label]
        
!T[table-label](a table)

|asda|asdasd|
|----|------|
|asda|asdasd|

",
      "type": "File",
    },
    "pos": Object {
      "end": 51,
      "start": 1,
    },
    "type": "Paragraph",
  },
  Object {
    "header": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 58,
                  "start": 54,
                },
                "text": "asda",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 58,
              "start": 54,
            },
            "type": "TableCell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 65,
                  "start": 59,
                },
                "text": "asdasd",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 65,
              "start": 59,
            },
            "type": "TableCell",
          },
        ],
        "parent": Object {
          "header": [Circular],
          "parent": Object {
            "children": [Circular],
            "parent": null,
            "path": "test.md",
            "pos": Object {
              "end": 96,
              "start": 0,
            },
            "raw": "
!TK[table-label]
        
!T[table-label](a table)

|asda|asdasd|
|----|------|
|asda|asdasd|

",
            "type": "File",
          },
          "pos": Object {
            "end": 94,
            "start": 53,
          },
          "rows": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "pos": Object {
                        "end": 86,
                        "start": 82,
                      },
                      "text": "asda",
                      "type": "Text",
                    },
                  ],
                  "parent": [Circular],
                  "pos": Object {
                    "end": 86,
                    "start": 82,
                  },
                  "type": "TableCell",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "pos": Object {
                        "end": 93,
                        "start": 87,
                      },
                      "text": "asdasd",
                      "type": "Text",
                    },
                  ],
                  "parent": [Circular],
                  "pos": Object {
                    "end": 93,
                    "start": 87,
                  },
                  "type": "TableCell",
                },
              ],
              "parent": [Circular],
              "pos": Object {
                "end": 94,
                "start": 81,
              },
              "type": "TableRow",
            },
          ],
          "type": "Table",
        },
        "pos": Object {
          "end": 66,
          "start": 53,
        },
        "type": "TableRow",
      },
      Object {
        "children": Array [
          Object {
            "align": "Default",
            "joinColsRight": undefined,
            "joinRowsUp": undefined,
            "parent": [Circular],
            "pos": Object {
              "end": 72,
              "start": 68,
            },
            "type": "TableControlCell",
          },
          Object {
            "align": "Default",
            "joinColsRight": undefined,
            "joinRowsUp": undefined,
            "parent": [Circular],
            "pos": Object {
              "end": 79,
              "start": 73,
            },
            "type": "TableControlCell",
          },
        ],
        "parent": Object {
          "header": [Circular],
          "parent": Object {
            "children": [Circular],
            "parent": null,
            "path": "test.md",
            "pos": Object {
              "end": 96,
              "start": 0,
            },
            "raw": "
!TK[table-label]
        
!T[table-label](a table)

|asda|asdasd|
|----|------|
|asda|asdasd|

",
            "type": "File",
          },
          "pos": Object {
            "end": 94,
            "start": 53,
          },
          "rows": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "pos": Object {
                        "end": 86,
                        "start": 82,
                      },
                      "text": "asda",
                      "type": "Text",
                    },
                  ],
                  "parent": [Circular],
                  "pos": Object {
                    "end": 86,
                    "start": 82,
                  },
                  "type": "TableCell",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "pos": Object {
                        "end": 93,
                        "start": 87,
                      },
                      "text": "asdasd",
                      "type": "Text",
                    },
                  ],
                  "parent": [Circular],
                  "pos": Object {
                    "end": 93,
                    "start": 87,
                  },
                  "type": "TableCell",
                },
              ],
              "parent": [Circular],
              "pos": Object {
                "end": 94,
                "start": 81,
              },
              "type": "TableRow",
            },
          ],
          "type": "Table",
        },
        "pos": Object {
          "end": 80,
          "start": 67,
        },
        "type": "TableControlRow",
      },
    ],
    "label": "table-label",
    "name": Array [
      Object {
        "children": Array [],
        "parent": Object {
          "keyArgs": Object {},
          "label": "table-label",
          "opcode": "T",
          "parent": Object {
            "children": Array [
              Object {
                "parent": [Circular],
                "pos": Object {
                  "end": 17,
                  "start": 1,
                },
                "text": "0",
                "type": "Text",
              },
              Object {
                "parent": [Circular],
                "pos": Object {
                  "end": 26,
                  "start": 18,
                },
                "type": "Br",
              },
            ],
            "parent": Object {
              "children": [Circular],
              "parent": null,
              "path": "test.md",
              "pos": Object {
                "end": 96,
                "start": 0,
              },
              "raw": "
!TK[table-label]
        
!T[table-label](a table)

|asda|asdasd|
|----|------|
|asda|asdasd|

",
              "type": "File",
            },
            "pos": Object {
              "end": 51,
              "start": 1,
            },
            "type": "Paragraph",
          },
          "pos": Object {
            "end": 51,
            "start": 27,
          },
          "posArgs": Array [
            Array [
              [Circular],
            ],
          ],
          "type": "OpCode",
        },
        "pos": Object {
          "end": 50,
          "start": 43,
        },
        "text": "a table",
        "type": "Text",
      },
    ],
    "parent": Object {
      "children": [Circular],
      "parent": null,
      "path": "test.md",
      "pos": Object {
        "end": 96,
        "start": 0,
      },
      "raw": "
!TK[table-label]
        
!T[table-label](a table)

|asda|asdasd|
|----|------|
|asda|asdasd|

",
      "type": "File",
    },
    "pos": Object {
      "end": 94,
      "start": 53,
    },
    "rows": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 86,
                  "start": 82,
                },
                "text": "asda",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 86,
              "start": 82,
            },
            "type": "TableCell",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "parent": [Circular],
                "pos": Object {
                  "end": 93,
                  "start": 87,
                },
                "text": "asdasd",
                "type": "Text",
              },
            ],
            "parent": [Circular],
            "pos": Object {
              "end": 93,
              "start": 87,
            },
            "type": "TableCell",
          },
        ],
        "parent": Object {
          "header": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "pos": Object {
                        "end": 58,
                        "start": 54,
                      },
                      "text": "asda",
                      "type": "Text",
                    },
                  ],
                  "parent": [Circular],
                  "pos": Object {
                    "end": 58,
                    "start": 54,
                  },
                  "type": "TableCell",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "pos": Object {
                        "end": 65,
                        "start": 59,
                      },
                      "text": "asdasd",
                      "type": "Text",
                    },
                  ],
                  "parent": [Circular],
                  "pos": Object {
                    "end": 65,
                    "start": 59,
                  },
                  "type": "TableCell",
                },
              ],
              "parent": [Circular],
              "pos": Object {
                "end": 66,
                "start": 53,
              },
              "type": "TableRow",
            },
            Object {
              "children": Array [
                Object {
                  "align": "Default",
                  "joinColsRight": undefined,
                  "joinRowsUp": undefined,
                  "parent": [Circular],
                  "pos": Object {
                    "end": 72,
                    "start": 68,
                  },
                  "type": "TableControlCell",
                },
                Object {
                  "align": "Default",
                  "joinColsRight": undefined,
                  "joinRowsUp": undefined,
                  "parent": [Circular],
                  "pos": Object {
                    "end": 79,
                    "start": 73,
                  },
                  "type": "TableControlCell",
                },
              ],
              "parent": [Circular],
              "pos": Object {
                "end": 80,
                "start": 67,
              },
              "type": "TableControlRow",
            },
          ],
          "parent": Object {
            "children": [Circular],
            "parent": null,
            "path": "test.md",
            "pos": Object {
              "end": 96,
              "start": 0,
            },
            "raw": "
!TK[table-label]
        
!T[table-label](a table)

|asda|asdasd|
|----|------|
|asda|asdasd|

",
            "type": "File",
          },
          "pos": Object {
            "end": 94,
            "start": 53,
          },
          "rows": [Circular],
          "type": "Table",
        },
        "pos": Object {
          "end": 94,
          "start": 81,
        },
        "type": "TableRow",
      },
    ],
    "type": "TableProcessed",
  },
]
`;
